pkg/apis/logstash/v1alpha1/logstash_types.go

```diff
@@ -31,6 +35,10 @@ type LogstashSpec struct {
 	// +kubebuilder:validation:Optional
 	Image string `json:"image,omitempty"`
 
+	// ElasticsearchRef is a reference to an Elasticsearch cluster running in the same Kubernetes cluster.
+	// +kubebuilder:validation:Optional
+	ElasticsearchRefs []commonv1.ObjectSelector `json:"elasticsearchRefs,omitempty"`
```

Comment:
Thinking forward a little to when we will have named clusters in Logstash, we talked about introducing the notion of a `clusterName` into the ElasticsearchRef definition along the lines of

```
elasticsearchRefs:
  - clusterName: k8sData
    name: local
    namespace: production               
```

with the `clusterName` being used as the id in the `cluster.yaml` equivalent that we envisage for Logstash.
If we think about a `cluster.yaml` looking something like:

```
elasticsearch_clusters:
  k8sData:
    host: "https://metaldata-es-http:9200"
    username: “logstash_system
    password: REDACTED
    ca_file: /mnt/elastic-internal/es-certs/ca.crt
```

I think we could do this by creating the `LogstashSpec` to look like:

```
type LogstashSpec struct {
  :
  // Named list of Elasticsearch Clusters
  ElasticsearchRefs[] ElasticsearchCluster `json:"elasticsearchRefs,omitempty"` 
  :
}

// Definition of type to define an Elasticsearch Cluster to be referred to in 
// Logstash pipeline
type ElasticsearchCluster struct {
  commonv1.ObjectSelector `json:",omitempty,inline"`
  ClusterName              string `json:"clusterName,omitempty"`
}
```

We could potentially create the env variables with the clusterName? (And if we did would we need the namespace here?)

There is some prior art with the elastic agent using an [outputName](https://github.com/elastic/cloud-on-k8s/blob/main/config/recipes/elastic-agent/multi-output.yaml#L8) to distinguish between multiple `elasticsearchRefs`.

Alternatively, we could stick with what we have here, and write `cluster.yaml` as:

```
elasticsearch_clusters:
  local_production:
    host: "https://metaldata-es-http:9200"
    username: “logstash_system
    password: REDACTED
    ca_file: /mnt/elastic-internal/es-certs/ca.crt
```

using `$namespace_$name` as the id.

WDYT?

Reply:
These are awesome suggestions. +1 to follow the prior art of agent to unify the experience. Also, +1 to bring cluster name in technical preview. 

I think we can use `outputName` instead of `clusterName` to further share the experience. For the env variable, let's replace `$namespace_$name` with `outputName`. The name needs to be unique in `elasticsearchRefs`. I will add validation.

Reply:
My thought on using `clusterName` over `outputName` is that we also use want to use `elasticsearchRef` in `input` and `filter` plugins, not just for outputs

Reply:
@robbavey Revisiting the usage of `clusterName`. I have doubts about its value. One drawback is the stack monitoring should understand `clusterName` and send metrics to the corresponding cluster, which does not feasible now because the operator only considers namespace + name in the association. 

From the design doc, things we want to achieve (now and future `cluster.yml`) with clusterName seem can be replaced by namespace + name.
```
    elasticsearch {
      cluster_id => namespace_name
    }
```
```
elasticsearch_clusters:
  namespace_name:
    host: "https://metaldata-es-http:9200"
    username: “logstash_system
    password: REDACTED
    ca_file: /mnt/elastic-internal/es-certs/ca.crt
  namespace_name2:
    cloud_id: XXXXXXX
     api_key: yyyyyy
```
`clusterName` works like a nickname that is handy when user wants to give a second name.
I wanna keep the current env variable naming convention, namespace + name, at least in the technical preview.

Reply:
Does stack monitoring need `clusterName`? Or can it just use the same settings that it has today? It is metricbeat sending the data, so I don't think it would need any changes? 

But, it does seem a little confusing having two different ways to specify `ElasticsearchRef` depending on whether it is a plugin usage or a stack monitoring usage (plus we'll need to consider central management soon)

I'm ok with keeping it as `namespace_name` for the technical preview. My main concern was that it was a little unwieldy, especially taking into account that users need to remember to change all their `-` to `_` when reading ENV variable, but if it turns out that it is a significant usability concern we can amend later.

Reply:
stack monitoring doesn't need `clusterName` to work, and it is perfectly fine to work the same way as of today with namespace + name. But when users give a nickname to a cluster, it is not nice that stack monitoring doesn't understand/ support nickname.

Reply:
ok - Let's keep it with `namespace_name` for technical preview, and change if it looks problematic

Reply:
What do you think about removing `DEFAULT_` if we are using the default namespace? And only including namespace if it is specified?

Reply:
k8s best practice is to not create any resources to the `default` namespace. I would expect users always create their namespaces, and keep `DEFAULT_` for consistency in case it is used.  An alternative could be creating two sets of env variables for the default namespace, one with `DEFAULT_` and one without a namespace.

Not sure if we will use `clusterName` or `namespace_name` in `cluster.yml` in the future. Having two sets of names may be confusing. I tend to believe users have knowledge of the name of the default namespace.

Reply:
I was trying to understand why the proposal of using the `clusterName` has been ruled out. The main reason is that concatenating the namespace with the name seems a bit brittle, in the sense that it does not really prevent collisions (yeah I know, it's a bit "far-fetched" 😄 ), and also because I think this is the kind of situation where, as a user, I would favor an explicit naming over an implicit one. It allows for example to change the target cluster without updating the configuration.

> One drawback is the stack monitoring should understand clusterName and send metrics to the corresponding cluster, which does not feasible now because the operator only considers namespace + name in the association.

I think it makes sense to treat stack monitoring as a special case because the concept of `clusterName` has no meaning in that context?

Reply:
I just realized that this would also allow users to install the sample manifest into a new namespace without needing to update the configuration:

```
k apply -f config/samples/logstash/logstash_es.yaml -n foo
```

```
[2023-04-20T13:58:05,513][ERROR][logstash.agent           ] An exception happened when converging configuration {:exception=>RuntimeError, :message=>"Could not fetch the configuration, message: Cannot evaluate `${DEFAULT_ELASTICSEARCH_SAMPLE_ES_HOSTS}`. Replacement variable `DEFAULT_ELASTICSEARCH_SAMPLE_ES_HOSTS` is not defined in a Logstash secret store or as an Environment entry and there is no default value given."}
```

Reply:
My main concern about not using `clusterName` is as a user, I am happy to give an explicit name, but it is very confusing that I cannot use the name in stack monitoring. As a user, I could not understand why stack monitoring is a special case.

Reply:
I tend to agree with @barkbay that the lack of `clusterName` in `elasticsearchRef` in stack monitoring is probably ok - `clusterName` is added as an aid for the user to refer to an elasticsearch cluster pointed to by `elasticsearchRef` outside of the kubernetes CRD in pipeline configurations, without needing to deal with kubernetes semantics, such as namespaces. The ability to define/move namespaces outside of the CRD is an an interesting twist on this.

Reply:
Could this be an optional feature? If users define it it will be used to name the environment variables? Also could it then just be named `alias`? 

```yaml
elasticsearchRefs:
  - alias: o11y-cluster
    name: monitor
    namespace: production  
```
 I think I also agree that monitoring is different enough to allow a special case here. Because users have to understand the relationship between the `elasticsearchRef` and what they define in configuration there is a motivation for such a construct. Why would one want to use an alias for stack monitoring?

Reply:
In Logstash context, `clusterName` is a mapping of name to endpoint (host, username, password, cacert). The endpoint can be an input (source) or an output (destination). It will be referenced in pipeline.

```
input { 
  kafka { } 
}
output {
  elasticsearch { cluster_name => "production" }
  elasticsearch { cluster_name => "uat" }
}
```

In Logstash CRD, we allow users to define the name mapping
```
  elasticsearchRefs:
    - clusterName: production
      name: es1
    - clusterName: uat
       name:es2
    - clusterName: monitor
       name:es3
```

However, Stack Monitoring does not understand the name `production`, `uat` and `monitor`. If Stack Monitoring wants to insert metrics to `monitor`, it needs the following

```
  monitoring:
    metrics:
      elasticsearchRefs:
        - name: es3
```
As a user, I am confused why I can't use `monitor` that I declared earlier. The name works in pipeline but not in Stack Monitoring.

IMO, ultimately we will need `clusterName` to achieve `cluster.yml` which contains the mapping of alias to endpoint. In technical preview, it seems not ready. I hope the example can explain the confusion.

Reply:
I guess my confusion here, is what would be your expectation of using `clusterName` in the `monitoring` section? 

Are you expecting to be able to use `clusterName` instead of `name` + `namespace`? Or as well as? Is your expectation that you could use 
```
 monitoring:
    metrics:
      elasticsearchRefs:
        - clusterName: monitor
```

or that you would expect 
```
 monitoring:
    metrics:
      elasticsearchRefs:
         - clusterName: monitor
            name:es3
```
or that `clusterName` is arbitrary without the specific `cluster` or `clusterName` setting in `cluster.yml` when that comes to be?

I agree that it feels a little half baked without the `cluster.yml`, and we only refer to `clusterName` in the env variable, but I do feel that it will have utility when we start allowing named clusters, and maybe introducing the notion earlier is of some value.

Thoughts @flexitrev, @jsvd, @roaksoax?

Reply:
my expectation is to be able to use `clusterName` to output metrics to `monitor`
```
 monitoring:
    metrics:
      elasticsearchRefs:
        - clusterName: monitor
```
and won't stop user to use namespace + name to output
```
  monitoring:
    metrics:
      elasticsearchRefs:
        - name: es3
```

The declaration in `elasticsearchRefs` will all convert to `cluster.yml`. Only the `clusterName` in `cluster.yml` can be referenced in pipeline.

```
  elasticsearchRefs:
    - clusterName: production
      name: es1
    - clusterName: uat
       name:es2
    - clusterName: monitor
       name:es3
```
cluster.yml
```
- production:
    host:...
    username:...
    password:...
    cacert:...
- uat:
    host:...
    username:...
    password:...
    cacert:...
- monitor:
    host:...
    username:...
    password:...
    cacert:...
```